Add CCE system equations

moxcodes opened the pull request:
## Proposed changes

This PR adds the Cauchy characteristic extraction system equations, which compute the set of inputs to the hypersurface integration routine, and represent the core of the physics at work here. 

The equations I've implemented are complicated and have been adjusted to the computational task from the equations in the literature. If any of the reviewers want to be thorough and check the form of the equations themselves (and not simply trust the forms in the documentation), please just leave a comment to that effect and I will create a repository which contains my notes and Mathematica document to produce these results. I would not recommend any reviewer investing the time needed to verify the form of the equations without those reference materials.

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #1398

moxcodes committed:
Add CCE equations

- add tags for relevant CCE quantities
- add computation methods for CCE integrands

moxcodes committed:
Add CCE Tags

nilsdeppe commented on tests/Unit/Pypp/PyppFundamentals.hpp:
```diff
@@ -212,6 +212,32 @@ struct ToPyObject<DataVector, std::nullptr_t> {
   }
 };
 
+template <>
```
I'd recommend factoring this commit out since it'll be easier to get in and should have different reviewers

moxcodes replied:
That makes good sense :+1:. I was struggling with whether I should ask @osheamonn for this one because of this change.

hrueter commented on docs/References.bib:
```diff
@@ -169,6 +169,37 @@ @book{Hartle2003gravity
+  eprint         = "gr-qc/9708065",
+  archivePrefix  = "arXiv",
+  primaryClass   = "gr-qc",
+  url            = {https://arxiv.org/abs/gr-qc/9708065}
```
use doi address.

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+  static std::string name() noexcept { return "Beta"; }
+};
+
+/// Bondi parameter \f$J\f$
```
J -> H

What is H actually? We can't find it in the Handmer paper.

moxcodes replied:
Ah, apparently notation has diverged for this quantity, and the Handmer paper doesn't discuss the procedure in detail. I'll add more explanatory detail to this comment :+1:.

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+};
+
+/// Bondi parameter \f$J\f$
+struct H : db::SimpleTag {
```
We suggest naming these `BondiH`, `BondiJ`, ...

moxcodes replied:
I have a mild preference for keeping the tags brief, due to the effect of expanding the already very long tag lists in `Equations.hpp`. The names are consistently used in CCE, and anything that accesses these tags outside the CCE module will need to access them by `Cce::Tags::H` etc, so I think it is sufficiently disambiguated.

nilsvu replied:
I’d still prefer `BondiH`, `BondiJ` etc, just to avoid single-letter type names that have the possibility of being hard to search-and-replace. I’d also like the other reviewers @markscheel @nilsdeppe to weigh in here before you make the change.

nilsdeppe replied:
Yea, I hate single letter things. With tags I don't feel as strongly because they're always prefixed with `Tags::`, but I still am not a fan of it. Since currently there apparently is a comment "Bondi parameter tags` necessary, @nilsleiffischer's suggestion also removes the need for that comment.

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+/// A prefix tag representing a quantity that will appear on the right-hand side
+/// of an explicitly regular differential equation
+template <typename Tag>
+struct IntegrandFor : db::PrefixTag, db::SimpleTag {
```
remove `For`. This is implied by the prefixing.

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+/// represent some of the cached quantities for CCE. Note that the spins are
+/// added.
+template <typename LhsTag, typename RhsTag>
+struct Times : db::PrefixTag, db::SimpleTag {
```
better name would be `Product`

moxcodes replied:
I will make the alteration to `Multiplies` instead by analogy to the stl

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+/// represent some of the cached quantities for CCE. Note that the spins are
+/// added.
+template <typename LhsTag, typename RhsTag>
+struct Times : db::PrefixTag, db::SimpleTag {
```
This looks like being general to spin weighted quantities. So better move to DataStructures/Tags.hpp or DataStructures/SpinWeighted.hpp?

fmahebert replied:
Or perhaps wait until this is needed elsewhere?

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+/// Temporary scalars for particular spin weights.
+template <size_t index, int spin>
+struct TemporarySpinWeightedScalar : db::SimpleTag {
+  using type = Scalar<SpinWeighted<ComplexDataVector, spin>>;
```
We propose to add a spinweighted prefix tag to DataStructures/Tags.hpp, that swaps the type of a tensor with a spinweighted type. Then you can use the `TempBuffer` with the aliases from Datastructures/Variables.hpp.

hrueter commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+};
+
+// Additional values which are frequently used in CCE calculations, and
+// therefore worth caching
```
Should these go into a `detail` namespace? Maybe in a separate file?

nilsdeppe replied:
I don't think we've put tags into `detail` namespaces very often (ever?). My issue with moving these into a `detail` namespace is that they then won't show up in Doxygen and nobody will be able to find them without reading the code.

hrueter reviewed (dismissed):
group review by @nilsleiffischer (main), @hrueter 
We did not yet review the full PR.

If you think it would be hard to follow the calculation in the documentation please either make the documentation self-contained or add your notes as TeX files to the /docs. They are not yet automatically built when building the documentation. If you think it makes sense to add provide the Mathematica notebooks in the documentation, we propose to publish them in a separate repository. Assign them a DOI and link them from our documentation. (use e.g. [zenodo](https://zenodo.org) )

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1153 @@
+  template <typename BondiAndSWCacheTagList, typename DerivativeTagList,
+            typename IntegrandTagList, typename SecondaryTagList,
+            typename TemporaryTagList>
+  void evaluate(
```
As discussed at the call yesterday, a comment of ours for some reason didn't make it into the review. We suggested to make this conform to the existing `db::mutate_apply` mechanism, see also #1413 (that PR only makes that mechanism easier to use, it works already fine right now). Let's talk in a hangout if anything is unclear.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+// Distributed under the MIT License.  See LICENSE.txt for details.
```
Should be two (2) lines

... how did this even get committed?

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+};
+
+/// Bondi parameter \f$K^2 = 1 + J \bar{J}\f$
+struct KSquared : db::SimpleTag {
```
@moxcodes notes to self: check if this is used.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+struct Dy : db::PrefixTag, db::SimpleTag {
+  using type = Scalar<SpinWeighted<ComplexDataVector, Tag::type::type::spin>>;
+  using tag = Tag;
+  static const size_t dir = 2;
```
- Please use a clearer name. Suggest: `dimension_to_differentiate`
- Perhaps open an issue to add documentation that an assumption is made about the (phi, theta, r) ordering of the CCE coordinates.

moxcodes replied:
Issue created and assigned to me.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+
+/// A prefix tag representing the contribution to a differential equation which
+/// has a contribution of a nontrivial linear operator in addition to the
+/// differential operator
```
For ease of reading, suggest: "A prefix tag representing a linear operator that acts on `Tag`"

Also include the equation in the documentation (here? elsewhere?), because the tag is specialized to a particular differential equation (note, e.g., the hardcoded spin).

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+
+/// A prefix tag representing the contribution to a differential equation which
+/// has a contribution of a nontrivial linear operator acting on the conjugate
+/// of the quantity being solved for in addition to the differential operator
```
"A prefix tag representing a linear operator that acts on the conjugate of `Tag`"

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+/// has a contribution of a nontrivial linear operator in addition to the
+/// differential operator
+template <typename Tag>
+struct LinearOperatorFor : db::PrefixTag, db::SimpleTag {
```
Reconsider name to better convey that this returns L, not LH.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+  using type = Scalar<SpinWeighted<ComplexDataVector, 0>>;
+  static std::string name() noexcept { return "DuR"; } };
+
+/// The value \f$\partial_y R / R\f$, where \f$R\f$ is the variable associated
```
Typo: `_y` -> `_u`

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+// Additional values which are frequently used in CCE calculations, and
+// therefore worth caching
+
+/// Coordinate value $(1 - y)$, which will be cached and sent to the
```
RIP, render these docs and fix missing `\f` in several places.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+};
+
+/// The value \f$\exp(2\beta)\f$, which is used frequently in the Bondi
+/// computations.
```
Drop second half of comment, it isn't very illuminating. Same on following tag.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,255 @@
+
+/// The value \f$ J (\bar{Q} - 2 \bar{\eth} \beta ) \f$, which must be computed
+/// as an intermediate step to CCE equations.
+struct JbarQMinus2EthBeta : db::SimpleTag {
```
Convention so far has been to use tag name `Xbar` for `\bar{X}`, so fix doxygen and/or tag name for consistency. Or rename to SadTag.

fmahebert commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1153 @@
+/// \brief A struct for providing a tmpl::list of integrand tags that need to be
+/// computed before integration can proceed for a given Bondi variable tag.
+template <typename BondiVariable>
+using integrands_to_compute_for_bondi_variable =
```
`integrands` -> `integrand_terms` (globally)

fmahebert commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1153 @@
+};
+template <>
+struct tags_needed_for_integrand_computation_impl<
+    Tags::IntegrandFor<Tags::Beta>, TagsCategory::Secondary> {
```
`Secondary` -> `IndependentOfIntegration` or similar

fmahebert commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1153 @@
+
+template <>
+struct tags_needed_for_integrand_computation_impl<
+    Tags::IntegrandFor<Tags::Beta>, TagsCategory::BondiAndSWCache> {
```
`BondiAndSWCache` -> `TagsToSwshDifferentiate`

fmahebert reviewed (changes requested):
Review with @markscheel (primary) and @moxcodes. We looked at the Tags.hpp and did a brief overview of the equations. Still more review needed.

nilsvu commented on docs/References.bib:
```diff
@@ -169,6 +169,37 @@ @book{Hartle2003gravity
   url        = "https://www.pearson.com/us/higher-education/program/Hartle-Gravity-An-Introduction-to-Einstein-s-General-Relativity/PGM298455.html",
 }
 
+@article{Bishop1997ik,
```
Please sort alphabetically (see guidelines linked at the top of this file)

nilsvu commented on src/DataStructures/Tags.hpp:
```diff
@@ -40,4 +41,15 @@ struct Modal : db::SimpleTag, db::PrefixTag {
 };
+
+template <typename Tag, int Spin>
+struct SpinWeighted : db::PrefixTag, db::SimpleTag {
```
Please document this tag

nilsvu commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,269 @@
+
+namespace Cce {
+// suppresses doxygen problems with these functions
+///\cond
```
add space `/// \cond`, also you don't need the explanatory comment, this is done at a lot of places in the code

moxcodes replied:
This one I felt should have a comment, as dox is usually not rendered for cpp, but there appears to be a small bug in doxygen tackling this file.

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1010 @@
+  // clang-tidy readability-avoid-const-params-in-decls, but the consts must
+  // stay
+  static void apply(
+      const gsl::not_null<SpinWeighted<ComplexDataVector, 0>*>  // NOLINT
```
I think you can remove the `const`s here, you only need them in the `cpp`.

moxcodes replied:
That is not true for this scenario, because the calling interface in this case needs to know that it is passing to a function with const arguments (otherwise it can't rely on the function it calls to not mutate its own const arguments). At any rate, I wanted to avoid the long list of nolint's and it does not compile without the const :P

nilsdeppe replied:
Modulo a compiler bug this is not how `const` works. You can safely remove the `const` on the `not_null` arguments in the forward declarations, there's no need to `NOLINT` anything. On the arguments that are references, the `const` matters. I tested this that it compiles the `Cce` and `Test_Cce` libraries clang7 and gcc6. We also do not `const` the `not_null` arguments everywhere else in the entire repository and have never had to `NOLINT` anything. We actually use `not_null` so we don't have to `NOLINT`.

moxcodes replied:
Fortunately, this also can be reverted to standard use now that the `Variables` interface has been dropped.

nilsvu commented on tests/Unit/Evolution/Systems/Cce/TestFunctions.py:
```diff
@@ -0,0 +1,195 @@
+# Distributed under the MIT License.
```
As of the guidelines introduced in #1382 please name the Python file the same as the cpp test file, without the `Test_` prefix.

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1010 @@
+ * `SpinWeighted<ComplexDataVector, N>`s for each of the Bondi arguments.
+ */
+template <typename IntegrandTag>
+struct ComputeBondiIntegrand;
```
Name this `ComputeBondiIntegrandImpl` and move into detail namespace

moxcodes replied:
I'm happy naming it Impl, but as the documentation for the individual calculations is tied to the specializations of this struct, it should remain in a rendered namespace, particularly for the density of TeX in that documentation.

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1010 @@
+ * `apply` function in the specialization of `ComputeBondiIntegrand`.
+ */
+template <typename IntegrandTag>
+struct MutateComputeBondiIntegrand {
```
name `ComputeBondiIntegrand`

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1010 @@
+ * - `temporary_cache` : `Temporary`
+ */
+template <typename IntegrandTag>
+struct ComputeBondiIntegrandFromVariables {
```
I suggest to move this into test helpers. If we end up needing this in the public interface we can always move it back, but it is not clear to me that we do.

moxcodes replied:
I'd still lean towards keeping at as part of the public interface, if only because the CCE/SWSH interfaces are currently working with the design goal of having interfaces which act on clusters of tags with some mathematical and functional commonality, in order to manage the combinatorically large number of values that appear in the calculations and impose some mild hierarchy. It ends up having a nice connection as well to the way the lists of tags are presented in this namespace, where if the tags were all put into a single large list, it would be more difficult to read and maintain.

If there is consensus that the Variables interface should not be included in `src`, I will move it to the tests.

Another alternative is to make the templates more general still, and permit them to take an arbitrary template, so this code is equally operable for sets of Variables or TaggedTuples.

nilsvu replied:
I'd like to get input from the other reviews on this. My intention here is to not make the interface unnecessarily complex before we know we need that functionality.

moxcodes replied:
Alright, I'm happy to have other input for this one.

My main point is just that while it's conceptually simple to just take a bunch of arguments, it will tend to be a frustrating interface to mechanically use for calling code for the large number of arguments for some of these.

@hrueter @fmahebert @markscheel opinions?

nilsvu replied:
Right, so I think we agree that calling these functions with a list of arguments is conceptually nice and clean, but cumbersome when one has to write it out. That is, however, why we have the struct with the `argument_tags` typelist, so that you can just hand the struct over to the DataBox and have it sort out the arguments for you. I suggest you look at the `GeneralizedHarmonic` code and tests for an example. Those are just structs with a plain `argument_tags` list and an `apply` function that takes those arguments.

When we discussed this last week you mentioned you are worried about having to call the functions explicitly in tests, which is why I suggested to move this "short-cut" code with the categories and `Variables` to the tests.

moxcodes replied:
From separate discussion, we concluded that it will be best to move these to the test helpers for this PR as suggested by @nilsleiffischer .

nilsvu reviewed (changes requested):
I think the move of the implementations into the cpp file worked out great 👍 A few more comments on the overall structure below. Also I think it would be good to "activate" the second reviewer, to get a second opinion and make sure we are converging to a state that will be merged. @markscheel would you review this?

nilsvu commented:
Oh @markscheel sorry I didn't realised you already reviewed this as part of the group review posted by @fmahebert

moxcodes commented:
Fixups posted for all points aside from moving the `Variables` to a test helper. The last point, I'll implement and post separately the chosen solution once the discussion is resolved.
Thanks for the reviews!

kidder commented:
Current status:

- [x] @hrueter and @nilsleiffischer to look at latest fix ups
- [ ] @markscheel and @fmahebert to look at latest fix ups

nilsvu commented on docs/References.bib:
```diff
@@ -200,6 +190,16 @@ @article{Handmer2014qha
   url            = "https://doi.org/10.1088/0264-9381/32/2/025008"
 }
 
+  @book{Hartle2003gravity,
```
remove the leading whitespace please

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -398,13 +399,13 @@ struct MutateComputeBondiIntegrand {
- * \brief An interface to ComputeBondiIntegrand which is convenient for passing
- * around clusters of bondi quantities in `Variables`
+ * \brief An interface to ComputeBondiIntegrandImpl which is convenient for
+ * passing around clusters of bondi quantities in `Variables`
```
Bondi

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -467,6 +468,12 @@ struct ComputeBondiIntegrandFromVariables {
+#ifndef __clang__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-but-set-parameter"
+#endif  // __clang__
```
I don't know if this is the best-practice way to do this, so perhaps another reviewer can confirm

moxcodes replied:
Do you mean specifically the `#ifndef __clang__` to select compiler? This was the method I was advised to use for a compiler-specific warning in #1145 , though the warning was ultimately removed by other changes, but I'm happy to entertain other methods.

nilsvu reviewed (changes requested):
Please squash the fixups and these minor additional changes

moxcodes commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1020 @@
+          IntegrandTag, TagsCategory::IndependentOfBondiIntegration>>;
+
+  template <typename... Args>
+  static void apply(Args... args) noexcept {
```
const & or &&

moxcodes commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1020 @@
+ * `SpinWeighted<ComplexDataVector, N>`s for each of the Bondi arguments.
+ */
+template <typename IntegrandTag>
+struct ComputeBondiIntegrandImpl;
```
The convenience impl should instead be a impl function moved in the ComputeBondiIntegrand function.

moxcodes commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,1020 @@
+
+// - Tags for Linear operator that acts on the conjugate of H
+template <>
+struct tags_needed_for_integrand_computation_impl<
```
These type aliases, for each term, should be a single struct, and the collections of tags should probably be placed in a separate file.

moxcodes commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,195 @@
+          using bondi_integrand_tag = typename decltype(y)::type;
+          using bondi_integrand_argument_list =
+              all_arguments_for_integrand<bondi_integrand_tag>;
+          // check the Variables interface
```
Once the Variables interface is in a test helpers, this comment should give a few more details as to why we are testing this way (i.e. to allow the test to use the equations without using a databox and without manually specifying each of the functions and each of the numerous arguments for each of those functions).

moxcodes reviewed (commented):
A summary of the conclusions from discussion with @nilsdeppe and @nilsleiffischer for the last few changes to these utilities.

moxcodes commented:
Fixups posted and rebased. The only remaining fixup commit is the one regarding the discussion about the Variables helper and the accompanying typedefs.

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,724 @@
+class ComplexDataVector;
+///\endcond
+
+/// Contains system-specific tools for Cauchy Characteristic Extraction.
```
What do you mean with system-specific here? Perhaps just “functionality” for CCE

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,724 @@
+ * Characteristic hypersurface equations.
+ *
+ * \details The template argument must be one of the integrand-related prefix
+ * tags applied to a Bondi quantity for which that integrand is required. The
```
The template argument is not the prefix tag itself, but a _prefixed_ tag, if I understand correctly

moxcodes replied:
I'll adjust the later part of the sentence to refer to 'prefix tags templated on a Bondi quantity tag' for improved clarity.

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,724 @@
+          Tags::Integrand<Tags::Beta>>,
+      swsh_derivative_tags_needed_for_integrand<Tags::Integrand<Tags::Beta>>,
+      integration_independent_tags_needed_for_integrand<
+          Tags::Integrand<Tags::Beta>>>;
```
To build these return and argument tags you now refer to the tags lists in the other file. From our discussion the other day I understood we’d go with another strategy, and if I misunderstood then I’d like to suggest it here:
- Build the tag lists self-contained within the specialization of `ComputeBondiIntegrand`. You can split it into `buffer_tags` or similar, possibly private, type lists to retain the categoristation, if you like.
- Remove the `TagsLists.hpp` from this PR, but keep the code around on another branch in case you’ll ever need it.
- Also remove the `integrand_terms_to_compute_for_bondi_variable` from this PR but keep the code on another branch. It is part of the code that does the computation, in my opinion, so add it in a PR that makes use of it.

Before you make this change, I’d like the other reviewers to confirm @markscheel @nilsdeppe, so that you don’t waste time

moxcodes replied:
Provided consensus, I'm fine doing the first two. I'd prefer, however, to keep the `integrand_terms_to_compute_for_bondi_variable`. It's actually used in the test, and without it being present here, I'd just have to reproduce the same set of types in the test either by direct listing inline or via type aliases there.

nilsvu commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,244 @@
+/// \f$L\f$ in an equation of the form,
+/// \f$ (y - 1) \partial_y H + L H + L^\prime \bar{H} = A + (1 - y) B \f$
+template <typename Tag>
+struct LinearFactor : db::PrefixTag, db::SimpleTag {
```
You use LinearFactor and LinearOperator interchangeably. I suggest to pick one. LinearOperator is actually quite general, so can you be more specific in naming this?

moxcodes replied:
I will revert all of them to 'linear factor'. I think the current version of the code should only have 'operator' in some comments. Ultimately, what I'm trying to communicate with this is a factor which acts in the precise way that is in the doxygen and `LinearFactor` seemed slightly better; do you have an alternative suggestion for a name for it?

nilsvu replied:
`LinearCoefficient` perhaps?

moxcodes replied:
After thinking about it, I disagree with `Coefficient`, as other parts of the code use it to refer to, for instance, spherical harmonic coefficients in a mode expansion.

nilsvu commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,244 @@
+  static std::string name() noexcept { return "Beta"; }
+};
+
+/// Bondi parameter \f$H = \partial_u J\f$. Note: the notation in the literature
```
\brief and \note

nilsvu commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,244 @@
+/// represent some of the cached quantities for CCE. Note that the spins are
+/// added.
+template <typename LhsTag, typename RhsTag>
+struct Multiplies : db::PrefixTag, db::SimpleTag {
```
I suggested to move this to SpinWeighted tags, since it’s not specific to CCE so shouldn’t be in this namespace. Did you disagree with the suggestion?

moxcodes replied:
Sorry, I incorrectly thought I had responded to this suggestion. While it is true that it isn't strictly specific to CCE, the more complicated assembly of tags that this is used for are less likely to be used generally. I'd prefer to keep it limited to the CCE scope until we find a need elsewhere.

nilsdeppe replied:
My concern with it being here is the code will be duplicated without your knowledge. We should put code where people expect it to be, not limit scope because we can't think of a use right now. Doing so is probably partly why SpEC at some point had 3 or more Levi-Civita iterator implementations.

moxcodes replied:
Alright, that's reasonable. I'll move it to `DataStructures/Tags.hpp`.

nilsvu commented on tests/Unit/DataStructures/Test_Tags.cpp:
```diff
@@ -59,9 +67,20 @@ void test_modal_tag() noexcept {
+      cpp17::is_same_v<
+          typename Tags::SpinWeighted<ComplexScalarTag, 1>::type,
+          Scalar<SpinWeighted<ComplexDataVector, 1>>>,
+      "Failed testing Tags::Modal<ScalarTag>");
```
SpinWeighted

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Equations.py:
```diff
@@ -0,0 +1,195 @@
+    jbar = np.conj(j)
+    k_squared = (1.0 + j * jbar)
+    qbar = np.conj(q)
+    script_av = eth_beta * eth_jbar + ethbar_ethbar_j / 2.0 \
```
You take this as an argument but then compute it

moxcodes replied:
It is a temporary buffer, taken by argument to avoid allocation.

nilsvu replied:
but this is in python, so does that actually work?

moxcodes replied:
oh, oops, I failed to note which file this was commenting on. No, it doesn't have any optimization utility in python. I could name the intermediate step something else, but I still need the argument so that the function signature is the same as the C++, and I still want to compute the intermediate step so that the function is reasonably readable.

nilsdeppe replied:
Maybe there is a way to comment out the name of the argument like there is in C/C++? Or maybe just name that argument something that makes it clear it's not used?

nilsvu replied:
You can use an underscore `_` for unused arguments, that's the convention

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,199 @@
+int assign_to_tag(
+    const gsl::not_null<Variables<VariablesTagList>*> all_variables,
+    const Arg arg) noexcept {
+  get(get<Tag>(*all_variables)).data() = arg;
```
Why do you need the `data()`? Looks dangerous :D

moxcodes replied:
This is a result of the SpinWeighted wrapper. The outer `get` obtains a `SpinWeighted<ComplexDataVector, N>`, and the `.data()` obtains the `ComplexDataVector`, which can be assigned regardless of the spin-weight. It's not a pointer, which I suspect is why it feels concerning.

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,199 @@
+template <typename Tag, typename VariablesTagList, typename Arg>
+int assign_to_tag(
+    const gsl::not_null<Variables<VariablesTagList>*> all_variables,
+    const Arg arg) noexcept {
```
const ref

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,199 @@
+    // create a variables which will hold everything
+    Variables<
+        tmpl::flatten<tmpl::append<ArgumentTagList, tmpl::list<OutputTag>>>>
+        all_variables{NumberOfGridPoints, std::complex<double>(10, 57)};
```
You can remove the second argument to construct with NaNs

nilsvu commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,724 @@
+          integrand_for_beta,
+      const SpinWeighted<ComplexDataVector, 2>& dy_j,        // NOLINT
+      const SpinWeighted<ComplexDataVector, 2>& j,           // NOLINT
+      const SpinWeighted<ComplexDataVector, 0>& one_minus_y  // NOLINT
```
You have to make sure that the apply function actually take the types of the return and argument tags. Since you want to avoid the many `get`s to bypass the `Scalar`s, you’ll need an `apply` and `apply_impl` or similar, I think @nilsdeppe suggested something.

moxcodes replied:
Ah, true. I agree this needs to be fixed.

nilsvu replied:
To make sure this works, I think it would be good to add a test that calls `db::mutate_apply`. This will be made easier with #1413

moxcodes committed:
Adjust Spin-Weighted Tag and struct for DataBox

- previously the Tags::SpinWeighted prefix tag and Tensors of spin-weighted
  quantities could not be easily put into a DataBox

moxcodes commented:
new fixups posted. I believe all comments have been addressed aside from expanding the single-character bondi tags to `BondiH` etc. pending input from @markscheel or @fmahebert . I want to request that @nilsdeppe at least reads the new commit for the compile-time Tensor get for `not_null` pointers, as it alters `Tensor.hpp`.

moxcodes commented:
squashed in a fix to get gcc to build

nilsvu commented:
@moxcodes I'll be on vacation 1st to 11th of May, so unable to review this in that time. As far as I understand this is still waiting on input by @markscheel, @fmahebert and @nilsdeppe though. If you end up moving this PR forward while I'm away, please feel free to replace me with another reviewer.

nilsdeppe commented:
@moxcodes could you please rebase and squash this, then ping me so I know to look at it? Thanks :)

moxcodes commented:
@nilsdeppe I've squashed the previous fixes and rebased. I've added one fixup in which I remove the `Variables` business. After the refactor of other utilities, I've seen the light and managed to use the rest of the cce tools in ways more consistent with the `DataBox` models.

nilsdeppe commented:
@moxcodes okay, would you mind squashing things again so I can look at the "final" proposed changes easily? :)

moxcodes commented:
@nilsdeppe Alright, squashed the fixes into their appropriate commits. 
Cheers

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -260,6 +260,22 @@ class Tensor<X, Symm, IndexList<Indices...>> {
       get(const Tensor<Args...>& t) noexcept;  // NOLINT
   // @}
 
+  // @{
```
This is really weird... I don't like it. No class ever does this. I haven't seen this pattern used in code outside of spectre and I think it unnecessarily complicates code by blurring the line between when an object is a pointer (and mutating) and when it is not. I would need to hear a really good argument as to why this should be in the code. I also generally dislike things like this because in my view they try to remove people too much from what the code actually means. It's like python, it's so "convenient" that you have no idea what the heck is happening and can't reason about anything anymore (okay, this isn't quite that bad, but it's a slippery slope). There are times when I think adding a layer is good, specifically when it hides a lot of unimportant details and saves a lot of code. I think in this case it does neither. It obscures important details and saves a character at best (since the resulting thing is again wrapper in `not_null` I'm not sure it saves any). Another issue is that with this `get(a)` would return a `not_null` and so the correct way to use it is `const auto b = get(a);`, whereas if `a` is not a `not_null` then it would be `const auto& b = get(a);`. While I'm sure you'd be willing to help police this, I'd rather not have to police it at all :)

Just to be clear, the long paragraph is just to explain my reasoning for generally disliking these sorts of short helpers that create ambiguities and don't make code easier to work with :)

moxcodes replied:
That is fine. I'm quite certain that this can be dropped from the PR without negative effect now that the `Variables` interface is gone.

nilsdeppe replied:
Yea, I only saw this being used in one place that I noticed :) I understand why you wanted it, but wanted to explain my experience why I think it'll be more trouble than fix problems. (I have a similar complaint about defaulting things because all too often I get that the default thing just doesn't do what I want and it takes me too long to figure out it was because of some hidden option) :)

nilsdeppe commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,244 @@
+/// \note the notation in the literature is not consistent regarding this
+/// quantity, or whether it is denoted by an \f$H\f$ at all. The SpECTRE CCE
+/// module consistently uses it to describe the (retarded) partial time
+/// derivative of \f$J\f$ at fixed numerical radius \f$y\f$.
```
I'm not sure what a `numerical radius` is as opposed to a non-numerical radius

moxcodes replied:
as opposed to Bondi radius which has physical meaning but is not used as a numerical coordinate system. Is something to that effect worth including in the comment?

nilsdeppe replied:
I think so, yes. Or maybe `compactified radius`, if that's still a true statement? I was just confused because to me any radius is "numerical" if I put it on a grid :)

nilsdeppe commented on src/DataStructures/Variables.hpp:
```diff
@@ -909,7 +909,7 @@ struct Subitems<TagList, Tag,
 namespace Tags {
 template <size_t N, typename T>
-struct TempTensor {
+struct TempTensor : db::SimpleTag {
```
Why this change? I don't consider something `Temp` if it is stored in a DataBox. In my opinion that's almost the definition of `not temporary`.

moxcodes replied:
I have a few buffers that are passed in to the computation functions to avoid allocations, and the computation functions are intended to act on a `DataBox`. Do you have an alternative suggestion for how to resolve that clash?

nilsdeppe replied:
I'll need to think about any possible alternatives a bit. Thank you for explaining the use case, that makes sense :)

nilsdeppe replied:
I can't find where this tag is used inside a DataBox in this PR. I'd be more comfortable with deferring this change to where it is actually necessary so I have more context to think about it. I also don't want this PR to be held up on this, since the rest looks good :)

moxcodes replied:
This is used via the type alias `TempScalar` in the tag lists for the quantities that need to be passed into the computation by `not_null` pointer. An example of one of those type lists is on `Equations.hpp:294`. This change is required to actually use those computational items via the `db::mutate_apply` in the tests.

nilsvu replied:
@nilsdeppe is this resolved for you?

nilsdeppe commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,196 @@
+  }
+};
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.Cce.Equations", "[Unit][Evolution]") {
```
I'll ask the bold question: are there no analytic solutions you can test this with? We've had a lot of bugs crop up because of bugs in random number tests.

moxcodes replied:
There exist analytic functions, but they require the other functionality for CCE before they would be at all useful (i.e. we would need all the infrastructure for actually taking all of these spin-weighted derivatives and ensuring that the relations between all of these inputs is correct). Also, the only known analytic functions that actually test the nonlinear parts are even more difficult, requiring an elliptic solve to generate. 
If you'd like me to open an issue for the analytic test, I'm happy to address it when there's more of the functionality in place.

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+    const SpinWeighted<ComplexDataVector, 2>& j,
+    const SpinWeighted<ComplexDataVector, 0>& one_minus_y) noexcept {
+  *integrand_for_beta =
+      one_minus_y / 8.0 *
```
A trick I learned is to never rely on the compiler for optimization: `/ 8.0` -> `* 0.125`, the reason is that unless the transformation is IEEE floating point compliant the compiler can't do it

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+                eth_j_jbar * (conj(j) * dy_j + j * conj(dy_j)) /
+                    (2.0 * (1.0 + j * conj(j))) -
+                (conj(j) * dy_j - j * conj(dy_j)) * eth_r_divided_by_r) /
+               4.0;
```
here too, `* 0.125`

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+    const SpinWeighted<ComplexDataVector, 1>& q,
+    const SpinWeighted<ComplexDataVector, 0>& k,
+    const SpinWeighted<ComplexDataVector, 0>& r) noexcept {
+  *regular_integrand_for_u = exp_2_beta / (2.0 * r) * (k * q - j * conj(q));
```
`0.5 * exp_2_beta ...`

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+       j * square(conj(eth_beta)) + j * conj(eth_eth_beta) +
+       eth_j_jbar * conj(eth_j_jbar) / (8.0 * k * (1.0 + j * conj(j))) +
+       0.5 *
+           (1.0 - eth_ethbar_j_jbar / 4.0 - eth_j_jbar * conj(eth_beta) -
```
`* 0.25`

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+       eth_j_jbar * conj(eth_j_jbar) / (8.0 * k * (1.0 + j * conj(j))) +
+       0.5 *
+           (1.0 - eth_ethbar_j_jbar / 4.0 - eth_j_jbar * conj(eth_beta) -
+            conj(ethbar_j) * ethbar_j / 2.0 - conj(j) * eth_ethbar_j / 2.0) /
```
`0.5 *` (always move the fundamental as far left as possible)

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+            conj(ethbar_j) * ethbar_j / 2.0 - conj(j) * eth_ethbar_j / 2.0) /
+           k +
+       k * (0.5 - eth_ethbar_beta - eth_beta * conj(eth_beta) -
+            q * conj(q) / 4.0) +
```
Here as well, `0.25 *` (I'll stop commenting, you got this :tada: :+1: )

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,883 @@
+// IWYU pragma: no_forward_declare Tags::SpinWeighted
+// IWYU pragma: no_forward_declare SpinWeighted
+
+///\cond
```
nitpick: please add spaces before the `\cond` when you squash (or post fixups) :)

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,883 @@
+};
+}  // namespace detail
+
+/// \brief A struct for providing a tmpl::list of integrand tags that need to be
```
ticks around `tmpl::list`

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,883 @@
+
+ private:
+  static void apply_impl(
+      const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*>  // NOLINT
```
here too

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,883 @@
+
+ private:
+  static void apply_impl(
+      const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*>  // NOLINT
```
here too

nilsdeppe commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,883 @@
+
+ private:
+  static void apply_impl(
+      const gsl::not_null<SpinWeighted<ComplexDataVector, 1>*>  // NOLINT
```
and here as well (I'm sure you can handle the rest of them)

moxcodes commented:
Fixups posted for all of the comments with clear conclusions. I dropped the commit with the `Tensor` `not_null` `get` overload.
I think the only outstanding comment after these fixups is the one regarding putting the temp tags in the `DataBox`.

nilsdeppe commented on src/DataStructures/Tags.hpp:
```diff
@@ -68,4 +69,25 @@ struct SpinWeighted : db::PrefixTag, db::SimpleTag {
+/// `detail::product_t` should be generalized to give a reasonable type for
+/// those cases, or template specializations should be constructed for this tag.
+template <typename LhsTag, typename RhsTag>
+struct Multiplies : db::PrefixTag, db::SimpleTag {
```
Could you please add a test for this tag?

nilsdeppe reviewed (commented):
Couple minor things. Fixups look good, I'm happy for you to squash them as well as the changes to my two new comments :)

moxcodes commented:
Fixups squashed in. Let me know if you have any more thoughts on the temporary tag @nilsdeppe.

nilsdeppe commented:
@nilsleiffischer @hrueter please give feedback on whether you're happy with the changes or not.

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+  static std::string name() noexcept { return "Ubar"; }
+};
+
+/// Bondi parameter \f$U\f$
```
U -> W ?

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+  }
+};
+
+// Below are additional values which are frequently used in CCE calculations,
```
Nitpick: these are tags associated with values (vs. simply values), no?

fmahebert commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+};
+
+/// The value \f$\partial_u R / R\f$, where \f$R\f$ is the variable associated
+/// with the shape of the initial worldtube data.
```
Could you provide the definition of R also in at least the Dy tag above?  Perhaps also for DuR immediately above.. you do say "worldtube parameter" there, but this phrase seems clearer to me.

fmahebert commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+
+void ComputeBondiIntegrand<Tags::LinearFactor<Tags::BondiH>>::apply_impl(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, 0>*>
+        linear_operator_for_h,
```
Tag is `LinearFactor<H>` but variable name is `linear_operator_for_h` ... is this intentional?

fmahebert commented on src/Evolution/Systems/Cce/Equations.cpp:
```diff
@@ -0,0 +1,268 @@
+void ComputeBondiIntegrand<Tags::LinearFactorForConjugate<Tags::BondiH>>::
+    apply_impl(
+        const gsl::not_null<SpinWeighted<ComplexDataVector, 4>*>
+            linear_operator_for_conjugate_h,
```
ditto

fmahebert commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,940 @@
+ * \bar{q}^B.\f]
+ * See \cite Bishop1997ik \cite Handmer2014qha for full details.
+ *
+ * in the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$
```
in -> In

fmahebert commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,940 @@
+};
+
+/*!
+ * \brief Computes the linear operator which multiplies \f$H\f$ in the
```
Here again, and a few places below in dox and in code: operator vs. factor?

fmahebert reviewed (commented):
Some comments re: documentation and consistency. I am happy for you to squash any fixes directly.

nilsdeppe commented:
@nilsleiffischer @hrueter please give feedback on whether you're happy with the changes or not.

nilsvu commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+// Bondi parameter tags
+
+/// Bondi parameter \f$\beta\f$
+struct Beta : db::SimpleTag {
```
For consistency with the other tags, this should probably also be `BondiBeta` (else I'd wonder what makes it special)

nilsvu commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+};
+
+/// \brief Bondi parameter \f$H = \partial_u J\f$.
+/// \note the notation in the literature is not consistent regarding this
```
`the` -> `The`

nilsvu commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+/// A prefix tag representing the boundary data for a quantity on the extraction
+/// surface.
+template <typename Tag>
+struct BoundaryValue : db::PrefixTag, db::SimpleTag {
```
You can remove the `Value` suffix. I leave it up to you to decide if you want to do this or not :)

moxcodes replied:
If you don't feel strongly about it, I think I'll leave it as is, if only because I'd like it to carry the connotation of being used during integral evaluations, for which I think it is more common to use the phrase 'the boundary value' for the definite integration.

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Equations.py:
```diff
@@ -0,0 +1,205 @@
+# Test function for beta equation
+
+
+def compute_integrand_for_beta(dy_j, j, one_minus_y):
```
Could you remove the `compute_` prefixes in this file? If they didn't compute things these functions would be pretty useless ;)

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Equations.py:
```diff
@@ -0,0 +1,205 @@
+           + one_minus_y**2 * dy_dy_j / (4.0 * r)
+
+
+def compute_linear_operator_for_h(script_dj, dy_j, j, one_minus_y):
```
I think this is `linear_factor` now (there may be more occurrences where you used `operator` synonymously with `factor` before making the naming consistent, I haven't checked)

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,202 @@
+          forward_to_pypp_with<ComputeBondiIntegrand<bondi_integrand_tag>,
+                               bondi_integrand_argument_list,
+                               bondi_integrand_tag, 5, ComplexDataVector>(
+              std::make_index_sequence<5>{},
```
Is the `5` the number of `all_bondi_tags`? Perhaps add a comment or even write as `tmp::size<all_bondi_tags>`.

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,202 @@
+        typename variables_tag::type{NumberOfGridPoints});
+    // populate it with the arguments passed in.
+    assign_to_tag_list(make_not_null(&box), ArgumentTagList{}, args...);
+    // the function takes sevaral variables to allow segmentation, but for the
```
`several`

Are these comments about `Variables` a remnant of your previous implementation? I don't quite see how they still apply - the `variables_tag` a few lines above has nothing to do with this as far as I can tell.

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Test_Equations.cpp:
```diff
@@ -0,0 +1,202 @@
+        ::Tags::Variables<tmpl::append<tmpl::list<OutputTag>, ArgumentTagList>>;
+
+    auto box = db::create<db::AddSimpleTags<variables_tag>>(
+        typename variables_tag::type{NumberOfGridPoints});
```
I don't think you actually need to put everything into the `variables_tag`, you could also construct `db::AddSimpleTags` from a parameter pack.

nilsvu reviewed (changes requested):
I'm quite happy about how this code is structured now :) thanks for your patience!

Only a few minor things:

moxcodes commented:
@nilsleiffischer @fmahebert I've posted fixups for the last few comments.
Cheers

fmahebert commented:
The fixups LGTM

nilsvu reviewed (commented):
This looks pretty much done to me. Sorry for not checking the fixups earlier, please rebase on develop and squash. Then, make sure Travis is happy and ping @sxs-collaboration/spectre-core-devs to get this finally merged :)

markscheel commented on src/DataStructures/Tags.hpp:
```diff
@@ -40,4 +42,52 @@ struct Modal : db::SimpleTag, db::PrefixTag {
+/// rise to compilation errors when used as `Tags::SpinWeighted<Tag, spin>`. If
+/// you find such trouble, consult the whitelist of possible Tensor storage
+/// types in `Tensor.hpp`.
+template <typename Tag, typename SpinConstant>
```
The second template argument is called `SpinConstant` here, but `Spin` in the above docs.

markscheel commented on src/Evolution/Systems/Cce/Tags.hpp:
```diff
@@ -0,0 +1,232 @@
+  static std::string name() noexcept { return "DuR"; }
+};
+
+/// The value \f$\partial_u R / R\f$, where \f$R\f$ is the variable associated
```
I would prefer R be defined everywhere as "the Bondi radius \f$R(u,\theta,\phi)\f$ of the worldtube" (which is what I think it is), which i think is much clearer than "the variable associated with the shape of the initial worldtube data".

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,940 @@
+ * \tfrac{1}{4} J \partial_y (\eth (\bar{J})) -  \tfrac{1}{4} \eth (\bar{J})
+ * \partial_y (J) + \frac{\eth (J \bar{J}) \partial_y (J \bar{J})}{8 K^2} -
+ * \frac{\bar{J} \eth (R) \partial_y (J)}{4 R}. \f]
+ * Due to the number of variables involved in CCE quantities, both the input and
```
The arguments below are not ptrs to `Variables`, so this comment should be removed/fixed.

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \f[(1 - y) \partial_y H + H + (1 - y) J (\mathcal{D}_J H +
+ * \bar{\mathcal{D}}_J \bar{H}) = A_J + (1 - y) B_J.\f]
+ * The quantity \f$1 + (1 - y) J \mathcal{D}_J\f$ is the linear factor
+ * pre-factor for the non-conjugated H, and is computed from the equation:
```
\f$H\$

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \bar{\mathcal{D}}_J \bar{H}) = A_J + (1 - y) B_J.\f]
+
+ * The quantity \f$ (1 - y) J \bar{\mathcal{D}}_J\f$ is the linear factor
+ * pre-factor for the non-conjugated H, and is computed from the equation:
```
\f$H\$

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * See \cite Bishop1997ik \cite Handmer2014qha for full details.
+ *
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$
+ * is determined by the worldtube initial data, the equation which
```
R is a function of u, correct?  So I would change 'worldtube initial data' to 'worldtube data'.  (Or even better would be to be more explicit, "...where \f$r(u,\theta,\phi)\f$ is the Bondi radius of the \f$y=\f$const surface, and \f$R(u,\theta,\phi)\f$ is the Bondi radius of the worldtube").

 Ditto in docs in all the other functions.

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$
+ * is determined by the worldtube initial data, the equation which
+ * determines \f$\beta\f$ on a surface of constant \f$u\f$ given \f$J\f$ on
+ * the same surface is, \f[\partial_y (\beta) = \frac{1}{8} (-1 + y)
```
nitpick: no comma after 'is'.

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$Q\f$ on a surface of constant \f$u\f$ given \f$J\f$ and \f$\beta\f$ on the
+ * same surface is written as,
```
nitpick: no comma after 'as'

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \f[(1 - y) \partial_y Q + 2 Q = A_Q + (1 - y) B_Q.\f]
+ * We refer to \f$A_Q\f$ as the "pole part" of the integrand and \f$B_Q\f$
+ * as the "regular part". The pole part is computed by this function, and has
+ * the expression,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$Q\f$ on a surface of constant \f$u\f$ given \f$J\f$ and \f$\beta\f$ on the
+ * same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \f[(1 - y) \partial_y Q + 2 Q = A_Q + (1 - y) B_Q. \f]
+ * We refer to \f$A_Q\f$ as the "pole part" of the integrand and \f$B_Q\f$ as
+ * the "regular part". The regular part is computed by this function, and has
+ * the expression,
```
nitpick: no comma here.

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$U\f$ on a surface of constant \f$u\f$ given \f$J\f$, \f$\beta\f$, and
+ * \f$Q\f$ on the same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$W\f$ on a surface of constant \f$u\f$ given \f$J\f$,\f$\beta\f$, \f$Q\f$,
+ * and \f$U\f$ on the same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \f[(1 - y) \partial_y W + 2 W = A_W + (1 - y) B_W.\f]
+ * We refer to \f$A_W\f$ as the "pole part" of the integrand and \f$B_W\f$
+ * as the "regular part". The pole part is computed by this function, and has
+ * the expression,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \f[(1 - y) \partial_y W + 2 W = A_W + (1 - y) B_W. \f]
+ * We refer to \f$A_W\f$ as the "pole part" of the integrand and \f$B_W\f$ as
+ * the "regular part". The regular part is computed by this function, and has
+ * the expression,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$W\f$ on a surface of constant \f$u\f$ given \f$J\f$, \f$\beta\f$, \f$Q\f$,
+ * \f$U\f$ on the same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \f[  B_W = \tfrac{1}{4} \partial_y (\eth (\bar{U})) + \tfrac{1}{4} \partial_y
+ * (\bar{\eth} (U)) -  \frac{1}{2 R} + \frac{e^{2 \beta} (\mathcal{A}_W +
+ * \bar{\mathcal{A}_W})}{4 R}, \f]
+ * where,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * in the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$W\f$ on a surface of constant \f$u\f$ given \f$J\f$,\f$\beta\f$, \f$Q\f$,
+ * \f$U\f$, and \f$W\f$ on the same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ *
+ * We refer to \f$A_J\f$ as the "pole part" of the integrand
+ * and \f$B_J\f$ as the "regular part". The pole part is computed by this
+ * function, and has the expression,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * \bar{H}) = A_J + (1 - y) B_J.\f]
+ * We refer to \f$A_J\f$ as the "pole part" of the integrand
+ * and \f$B_J\f$ as the "regular part". The pole part is computed by this
+ * function, and has the expression,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$W\f$ on a surface of constant \f$u\f$ given \f$J\f$,\f$\beta\f$, \f$Q\f$,
+ * \f$U\f$, and \f$W\f$ on the same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * + \frac{1}{2 R}\right)\bigg]\notag\\
+ * &+ (1 - y)^2 \bigg[ \frac{\partial_y (\partial_y (J)) }{4 R} \bigg],
+ * \f}
+ * where,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$W\f$ on a surface of constant \f$u\f$ given \f$J\f$,\f$\beta\f$, \f$Q\f$,
+ * \f$U\f$, and \f$W\f$ on the same surface is written as,
```
nitpick: no comma

markscheel commented on src/Evolution/Systems/Cce/Equations.hpp:
```diff
@@ -0,0 +1,944 @@
+ * In the compactified coordinate \f$ y \equiv 1 - 2 R/ r\f$, where \f$R\f$ is
+ * determined by the worldtube initial data, the equation which determines
+ * \f$W\f$ on a surface of constant \f$u\f$ given \f$J\f$,\f$\beta\f$, \f$Q\f$,
+ * \f$U\f$, and \f$W\f$ on the same surface is written as,
```
nitpick: no comma at the end

markscheel reviewed (changes requested):
All minor documentation changes.

I did not try to check the equations themselves (ugh); that's what the tests are for...

moxcodes commented:
@markscheel I've posted fixups for all of the suggestions.
Cheers!

moxcodes commented:
@markscheel Just waiting for the go-ahead before I squash the fixups.
Thanks!

markscheel commented:
please squash!

moxcodes commented:
squashed (and rebased)
Cheers!

kidder commented:
@hrueter @nilsleiffischer @fmahebert @markscheel this should be ready for final approval.  Please only request changes for obvious errors, anything else, just comment and @moxcodes can decide to address them or make an issue.

nilsvu reviewed (approved):
@moxcodes What I think is missing for these equations are tests that do not rely on the Python implementation, but work with some honest, hard-coded numbers that were computed externally. This is to make sure the C++ and Python implementation don't have the same bug, which can easily happen when coding them up. This was discussed briefly in this review before. Please consider adding such tests soon, perhaps make an issue for the record.

moxcodes commented:
@nilsleiffischer This pull request alone doesn't have enough of the CCE infrastructure to have useful comparisons against hand-computed quantities. I agree that those checks are important (some already exist on my own branch), so I've created issue #1641 as a reminder to prioritize those tests once the CCE code is capable of them.

moxcodes commented:
@sxs-collaboration/spectre-core-devs since @nilsleiffischer was the primary on the 'request changes' from @hrueter , I think this one is actually good to go?
